16-projects.Rmd

```diff
@@ -4,18 +4,17 @@ When you work on larger projects or reports, you may not want to put all text an
 
 ## Source external R scripts {#source-script}
 
-If your R Markdown document has a large amount of code, you may consider putting some code in external R scripts, and run these scripts via `source()`\index{source()} or `sys.source()`\index{sys.source()}, e.g.,
+If your R Markdown document has a large amount of code, you may consider putting some code in external R scripts, and running these scripts with `source()`\index{source()}[^sys-source]. This not only makes your R Markdown document cleaner, but can also make it more convenient to develop R code since (e.g.) debugging R code is often easier in pure R scripts.
 
 ````md
 ```{r, include=FALSE}`r ''`
-source("your-script.R", local = knitr::knit_global())
```

Comment:
But this is more robust than `local = TRUE`. It works anywhere (e.g. inside a function). It's an absolute environment, and `local = TRUE` is a relative environment, which is subject to your current working directory/environment, and could be confusing when it is not called directly in a code chunk but from another function, e.g.,

```r
load_scripts <- function(files) {
  for (f in files) source(f, local = TRUE)
}
```

Both `local = TRUE` and `FALSE` have their usefulness. There isn't an absolutely correct choice, so I didn't strongly emphasize one or another. Perhaps the distinction between the two choices should be explained more clearly in this section. In general, I feel `source("your-script.R", local = knitr::knit_global())` is still the safest choice.

Reply:
Yeah, if you want to run `source()` inside a function and have it impact code outside of the function, you're going to have to do something special. As a general principle, I don't think functions should mutate the global environment in this way, so I do believe that `local = TRUE` is the correct recommendation here. (And then mention `local = knitr::knitr_global()` as a special case if you _really_ want to do that)

Reply:
But this is now really about advice for writing functions that call source, not about calling `source()` in knitr chunks.